Fix bug in quota describe output

derekwaynecarr opened the pull request:
There are resource quota values that are just counting number of an item (services, pods, secrets, etc.).  When there are none of that particular item in the namespace, the quota describe output is confusing because we default to show a "m" suffix.  When there is 0 of something, its just zero and there is no needed to specify a unit for `DecimalSI` notation.  By removing the suffix, our quota describer makes more sense since most people will ask, what does 0m services mean.

BEFORE:

``` shell
cluster/kubectl.sh describe quota quota --namespace=quota-example
Name:           quota
Resource        Used        Hard
--------        ----        ----
persistentvolumeclaims  0m      10
services        0m      5
```

AFTER:

``` shell
cluster/kubectl.sh describe quota quota --namespace=quota-example
Name:           quota
Resource        Used        Hard
--------        ----        ----
persistentvolumeclaims  0       10
services        0       5
```

derekwaynecarr committed:
Improve quota describe output

derekwaynecarr commented:
So I am not sure if this is the proper fix, but I am still trying to work my way down to the root issue.

If I do the following simple test:

```
q := NewQuantity(int64(0), DecimalSI)
q.String() -> I get "0" with no units before this change
```

If I spit out the value created here in log:

https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/resourcequota/resource_quota_controller.go#L196

I get "0" and even after I set it in the resource list here:

https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/resourcequota/resource_quota_controller.go#L222

I get "0".

But if I GET the document back from the api server, I get "0m".

It makes me think there is something else going on in a deepCopy somewhere that is making a bad result occur.

The fix in this PR does correct the output, but I am not sure if its the best solution since its a bit of a mystery to me right now where the wrong conversion is happening.

Help appreciated if you are not happy with this PR, but will keep poking at it.

lavalamp commented:
Can you add a (failing) test case before you make this change? The tests for this are pretty extensive, it should be easy to stick one in somewhere. That will help me evaluate the change :) :)

derekwaynecarr commented:
@lavalamp - see my last comment, I am unable to produce a failing test in quantity_test.go since it works fine there... I am guessing something is happening as part of conversion somewhere now that is causing an unintended change in the meaning of 0.

derekwaynecarr commented:
@lavalamp - one theory is that generated deep copy conversion could be wrong:

https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/v1beta3/deep_copy_generated.go#L31

vs.

https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/resource/quantity.go#L373

derekwaynecarr commented:
Just looking at the code, I see some differences in copy behavior.  It looks like we set amount to nil in the deep_copy_generated code versus 0 in the quantity copy method.

lavalamp commented:
The autogenerated deep copy is possibly wrong, albiet not in a way that I expect to cause this. If that's it though, you should be able to make a test case that simulates its wrongness.

lavalamp commented:
If it is in fact the deep copy being wrong then we need to fix it, as it may have more onerous consequences for other numbers. Sorry for being pedantic about insisting on a test case.

derekwaynecarr commented:
@lavalamp - I want a test case too, I am trying to find where the best test case is needed since its not 100% clear where things are going astray... will update when I have more info.

derekwaynecarr commented:
Found the root issue:

https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/resource/quantity.go#L193

The call to amount.Round rounds even if the value is 0.

That makes the internal representation change from 0 to 0.000 when parsing "0" as input, and that ultimately causes the change in reported units.  Will get a unit test to demonstrate error in practice now.

derekwaynecarr commented:
@lavalamp - this should be set for review.

1) Added test case to quantity_test.go that shows the fixed issue with rounding on 0.
2) Added test case to conversion that demonstrates round-tripping the object from internal to external back to internal representation does not change the quantity value.

lavalamp commented on pkg/api/resource/quantity_test.go:
```diff
@@ -284,13 +292,19 @@ func TestQuantityString(t *testing.T) {
 	}
 	desired := &inf.Dec{} // Avoid modifying the values in the table.
 	for _, item := range table {
-		if item.in.Amount.Cmp(decZero) == 0 {
```
how about ```if item.expect == "0" {```-- it's a lot less of a change.

derekwaynecarr replied:
I am fine reverting this part of PR as it was not the core issue.  Apologies for missing it, but it did test what the generated code produced

lavalamp commented on pkg/api/resource/quantity.go:
```diff
@@ -190,7 +190,9 @@ func ParseQuantity(str string) (*Quantity, error) {
 	// Arguably, this should be inf.RoundHalfUp (normal rounding), but
 	// that would have the side effect of rounding values < .5m to zero.
-	amount.Round(amount, 3, inf.RoundUp)
+	if v, _ := amount.Unscaled(); v != int64(0) {
```
```if v, err := amount.Unscaled(); v != int64(0) || err != nil {```

lavalamp commented on pkg/api/resource/quantity_test.go:
```diff
@@ -57,6 +57,13 @@ func TestDec(t *testing.T) {
 
+func TestQuantityParseZero(t *testing.T) {
+	zero := MustParse("0")
+	if expected, actual := "0", zero.String(); expected != actual {
```
Did this test fail before your change?

derekwaynecarr replied:
Yes.  Each of the tests I sent failed before the change. 

Sent from my iPhone

> On Jun 2, 2015, at 4:27 PM, Daniel Smith <notifications@github.com> wrote:
> 
> In pkg/api/resource/quantity_test.go:
> 
> > @@ -57,6 +57,13 @@ func TestDec(t *testing.T) {
> >  	}
> >  }
> >  
> > +func TestQuantityParseZero(t *testing.T) {
> > +	zero := MustParse("0")
> > +	if expected, actual := "0", zero.String(); expected != actual {
> Did this test fail before your change?
> 
> —
> Reply to this email directly or view it on GitHub.
>

derekwaynecarr replied:
To be clear this and the conversion test is the failure tests fixed in this PR

lavalamp commented on pkg/api/resource/quantity_test.go:
```diff
@@ -275,6 +282,7 @@ func TestQuantityString(t *testing.T) {
 		{Quantity{dec(3, 3), DecimalExponent}, "3e3"},
 		{Quantity{dec(3, 3), DecimalSI}, "3k"},
 		{Quantity{dec(0, 0), DecimalExponent}, "0"},
+		{Quantity{Amount: nil, Format: DecimalSI}, "0"},
```
Did this test fail before your change?

derekwaynecarr replied:
Actually, this was an outlier as I tried to find the cause.  I can remove the test if you want but this is the result of what we get from our conversion functions so I think the test is still good

derekwaynecarr commented:
@lavalamp - reduced the PR to just the essential change in test cases, and updated code feedback.

lavalamp commented on pkg/api/resource/quantity.go:
```diff
@@ -190,7 +190,9 @@ func ParseQuantity(str string) (*Quantity, error) {
 	// Arguably, this should be inf.RoundHalfUp (normal rounding), but
 	// that would have the side effect of rounding values < .5m to zero.
-	amount.Round(amount, 3, inf.RoundUp)
+	if v, ok := amount.Unscaled(); v != int64(0) && ok {
```
I think you want ```|| !ok```-- you should do the rounding if the Unscaled value doesn't fit in an int64.

lavalamp replied:
Thinking about this more-- if you look down to line 266 ```exponent := int(-q.Amount.Scale())``` seems to be the culprit, if the value is 0 then the scale should be ignored.

I think this change here is good, but you should still be able to trigger the invalid behavior by constructing a value, so maybe you can fix the output, too?

(e.g., construct a value of "1000m", then use inf package to subtract 1000 from it and construct a new Quantity with that; this should trigger the incorrect output behavior.)

derekwaynecarr commented:
@lavalamp - made last requested change

lavalamp commented:
@derekwaynecarr I think our last messages crossed paths :)

derekwaynecarr commented:
@lavalamp - never met a big number library I liked... I am now wondering if our suffix generator should just always and forever provide no suffix if its 0 ;-)

derekwaynecarr commented:
@lavalamp - maybe 3rd times the charm, but I found the cleanest solution was to check explicitly for 0 in the Canonicalize() method after the nil Amount check using a simple Cmp.  That way, zero is always "0" on output ;-)

lavalamp commented:
LGTM, thanks-- sorry for being a PITA about this code :)

derekwaynecarr commented:
@lavalamp - I am glad we are fixed on both sides now, and libraries like these are pretty fundamental so its good to get it right.
